File: ios/SmashCore/Sources/SmashCore/Match/PlayerMotion.swift
/// How players move (spec §3, §8.1, §10): toward their target in play, easing to a stop outside
/// it, patrolling dummies on their rails, contact between players, and the boundary and nets.
extension Match {
    // MARK: §4.5 step 5

    mutating func movePlayers(_ dt: Double, live: Bool) {
        let blend = 1.0 - DetMath.exp(-(Tuning.Player.accelerationRate * dt))
        for i in players.indices {
            if players[i].pickupCooldown > 0 { players[i].pickupCooldown -= dt }
            if players[i].isDummy {
                movePatrol(i, dt)
                continue
            }
            var want = Vec.zero
            if live, let t = players[i].target {
                let dx = t.x - players[i].pos.x
                let dz = t.z - players[i].pos.z
                let d = Pitch.length(dx, dz)
                if d > Tuning.Player.arriveDeadZone {
                    let speed = Pitch.lesser(players[i].topSpeed, Tuning.Player.arriveGain * d)
                    want = Vec(x: dx / d * speed, z: dz / d * speed)
                }
            }
            var v = players[i].vel
            v.x += (want.x - v.x) * blend
            v.z += (want.z - v.z) * blend
            if !live {
                v.x *= Tuning.Match.easeFactor
                v.z *= Tuning.Match.easeFactor
            }
            players[i].vel = v
            players[i].pos = Vec(x: players[i].pos.x + v.x * dt, z: players[i].pos.z + v.z * dt)
            if players[i].isGoalie {
                players[i].facing = players[i].team == 0 ? 0 : Pitch.pi
            } else if Pitch.length(v.x, v.z) > Tuning.Player.turnMinSpeed {
                let turn = Pitch.angleDiff(Pitch.heading(v.x, v.z), players[i].facing)
                players[i].facing = Pitch.wrap(players[i].facing + turn * Pitch.lesser(1, Tuning.Player.turnRate * dt))
            }
        }
    }

    /// A dummy: a static one stands, a patrolling one slides between its two points (§10).
    mutating func movePatrol(_ i: Int, _ dt: Double) {
        guard let patrol = players[i].patrol else {
            players[i].vel = .zero
            return
        }
        let start = players[i].start
        let k = 0.5 + 0.5 * DetMath.sin(time * patrol.speed + patrol.phase)
        let next = Vec(x: start.x + (patrol.to.x - start.x) * k, z: start.z + (patrol.to.z - start.z) * k)
        if players[i].patrolFresh {
            players[i].vel = .zero
            players[i].patrolFresh = false
        } else {
            players[i].vel = Vec(x: (next.x - players[i].pos.x) / dt, z: (next.z - players[i].pos.z) / dt)
        }
        players[i].pos = next
    }

    // MARK: §4.5 step 6

    /// One pass over every pair in roster order, positions updated in place (§3).
    mutating func resolveContacts() {
        let n = players.count
        for i in 0..<n {
            for j in (i + 1)..<n {
                let dx = players[j].pos.x - players[i].pos.x
                let dz = players[j].pos.z - players[i].pos.z
                let d = Pitch.length(dx, dz)
                let reach = players[i].radius + players[j].radius
                guard d < reach && d > Tuning.Sim.contactEpsilon else { continue }
                let aFixed = players[i].isDummy
                let bFixed = players[j].isDummy
                if aFixed && bFixed { continue }
                let nx = dx / d
                let nz = dz / d
                let overlap = reach - d
                let pa = aFixed ? 0 : (bFixed ? overlap : overlap / 2)
                let pb = bFixed ? 0 : (aFixed ? overlap : overlap / 2)
                players[i].pos = Vec(x: players[i].pos.x - nx * pa, z: players[i].pos.z - nz * pa)
                players[j].pos = Vec(x: players[j].pos.x + nx * pb, z: players[j].pos.z + nz * pb)
                let closing = (players[j].vel.x - players[i].vel.x) * nx + (players[j].vel.z - players[i].vel.z) * nz
                if closing < 0 {
                    let half = closing * 0.5
                    if !aFixed { players[i].vel = Vec(x: players[i].vel.x + nx * half, z: players[i].vel.z + nz * half) }
                    if !bFixed { players[j].vel = Vec(x: players[j].vel.x - nx * half, z: players[j].vel.z - nz * half) }
                }
            }
        }
    }

    /// Everyone but a dummy is kept inside the boundary and out of both nets; they may walk
    /// round and behind a net (§3).
    mutating func constrainPlayers() {
        typealias P = Tuning.Pitch
        for i in players.indices where !players[i].isDummy {
            let r = players[i].radius
            let (sd, n) = Pitch.boundary(players[i].pos.x, players[i].pos.z, corner: sport.cornerRadius)
            if sd > -r {
                let pen = sd + r
                players[i].pos = Vec(x: players[i].pos.x - n.x * pen, z: players[i].pos.z - n.z * pen)
                let vn = players[i].vel.x * n.x + players[i].vel.z * n.z
                if vn > 0 { players[i].vel = Vec(x: players[i].vel.x - n.x * vn, z: players[i].vel.z - n.z * vn) }
            }
            for team in 0..<2 {
                let gz = Pitch.ownGoalZ(team)
                let dir = Pitch.direction(team)
                let back = gz - dir * P.goalDepth
                let zMin = Pitch.lesser(gz, back) - r
                let zMax = Pitch.greater(gz, back) + r
                let hx = P.goalMouthWidth / 2 + r
                let x = players[i].pos.x
                let z = players[i].pos.z
                guard x.magnitude < hx && z > zMin && z < zMax else { continue }
                let outLeft = x + hx
                let outRight = hx - x
                let outFront = dir > 0 ? zMax - z : z - zMin
                let outBack = dir > 0 ? z - zMin : zMax - z
                let least = Pitch.lesser(Pitch.lesser(outLeft, outRight), Pitch.lesser(outFront, outBack))
                if least == outLeft {
                    players[i].pos.x = -hx
                } else if least == outRight {
                    players[i].pos.x = hx
                } else if least == outFront {
                    players[i].pos.z = dir > 0 ? zMax : zMin
                } else {
                    players[i].pos.z = dir > 0 ? zMin : zMax
                }
            }
        }
    }
}

File: android/core/src/main/kotlin/in/nann/smashhockey/core/match/PlayerMotion.kt
package `in`.nann.smashhockey.core.match

import `in`.nann.smashhockey.core.generated.Tuning
import `in`.nann.smashhockey.core.math.DetMath
import kotlin.math.PI
import kotlin.math.abs

// How players move (spec §3, §8.1, §10): toward their target in play, easing to a stop outside
// it, patrolling dummies on their rails, contact between players, and the boundary and nets.

/** §4.5 step 5. */
internal fun Match.movePlayers(dt: Double, live: Boolean) {
    val blend = 1.0 - DetMath.exp(-(Tuning.Player.accelerationRate * dt))
    for (p in players) {
        if (p.pickupCooldown > 0) p.pickupCooldown -= dt
        if (p.isDummy) {
            movePatrol(p, dt)
            continue
        }
        var wantX = 0.0
        var wantZ = 0.0
        val t = p.target
        if (live && t != null) {
            val dx = t.x - p.pos.x
            val dz = t.z - p.pos.z
            val d = Pitch.length(dx, dz)
            if (d > Tuning.Player.arriveDeadZone) {
                val speed = Pitch.lesser(p.topSpeed, Tuning.Player.arriveGain * d)
                wantX = dx / d * speed
                wantZ = dz / d * speed
            }
        }
        var vx = p.vel.x
        var vz = p.vel.z
        vx += (wantX - vx) * blend
        vz += (wantZ - vz) * blend
        if (!live) {
            vx *= Tuning.Match.easeFactor
            vz *= Tuning.Match.easeFactor
        }
        p.vel = Vec(vx, vz)
        p.pos = Vec(p.pos.x + vx * dt, p.pos.z + vz * dt)
        if (p.isGoalie) {
            p.facing = if (p.team == 0) 0.0 else PI
        } else if (Pitch.length(vx, vz) > Tuning.Player.turnMinSpeed) {
            val turn = Pitch.angleDiff(Pitch.heading(vx, vz), p.facing)
            p.facing = Pitch.wrap(p.facing + turn * Pitch.lesser(1.0, Tuning.Player.turnRate * dt))
        }
    }
}

/** A dummy: a static one stands, a patrolling one slides between its two points (§10). */
private fun Match.movePatrol(p: Athlete, dt: Double) {
    val patrol = p.patrol
    if (patrol == null) {
        p.vel = Vec.ZERO
        return
    }
    val start = p.start
    val k = 0.5 + 0.5 * DetMath.sin(time * patrol.speed + patrol.phase)
    val next = Vec(start.x + (patrol.to.x - start.x) * k, start.z + (patrol.to.z - start.z) * k)
    if (p.patrolFresh) {
        p.vel = Vec.ZERO
        p.patrolFresh = false
    } else {
        p.vel = Vec((next.x - p.pos.x) / dt, (next.z - p.pos.z) / dt)
    }
    p.pos = next
}

/** §4.5 step 6: one pass over every pair in roster order, positions updated in place (§3). */
internal fun Match.resolveContacts() {
    val n = players.size
    for (i in 0 until n) {
        val a = players[i]
        for (j in i + 1 until n) {
            val b = players[j]
            val dx = b.pos.x - a.pos.x
            val dz = b.pos.z - a.pos.z
            val d = Pitch.length(dx, dz)
            val reach = a.radius + b.radius
            if (!(d < reach && d > Tuning.Sim.contactEpsilon)) continue
            val aFixed = a.isDummy
            val bFixed = b.isDummy
            if (aFixed && bFixed) continue
            val nx = dx / d
            val nz = dz / d
            val overlap = reach - d
            val pa = if (aFixed) 0.0 else if (bFixed) overlap else overlap / 2
            val pb = if (bFixed) 0.0 else if (aFixed) overlap else overlap / 2
            a.pos = Vec(a.pos.x - nx * pa, a.pos.z - nz * pa)
            b.pos = Vec(b.pos.x + nx * pb, b.pos.z + nz * pb)
            val closing = (b.vel.x - a.vel.x) * nx + (b.vel.z - a.vel.z) * nz
            if (closing < 0) {
                val half = closing * 0.5
                if (!aFixed) a.vel = Vec(a.vel.x + nx * half, a.vel.z + nz * half)
                if (!bFixed) b.vel = Vec(b.vel.x - nx * half, b.vel.z - nz * half)
            }
        }
    }
}

/** Everyone but a dummy is kept inside the boundary and out of both nets (§3). */
internal fun Match.constrainPlayers() {
    val pitch = Tuning.Pitch
    for (p in players) {
        if (p.isDummy) continue
        val r = p.radius
        val b = Pitch.boundary(p.pos.x, p.pos.z, sport.cornerRadius)
        if (b.distance > -r) {
            val n = b.normal
            val pen = b.distance + r
            p.pos = Vec(p.pos.x - n.x * pen, p.pos.z - n.z * pen)
            val vn = p.vel.x * n.x + p.vel.z * n.z
            if (vn > 0) p.vel = Vec(p.vel.x - n.x * vn, p.vel.z - n.z * vn)
        }
        for (team in 0 until 2) {
            val gz = Pitch.ownGoalZ(team)
            val dir = Pitch.direction(team)
            val back = gz - dir * pitch.goalDepth
            val zMin = Pitch.lesser(gz, back) - r
            val zMax = Pitch.greater(gz, back) + r
            val hx = pitch.goalMouthWidth / 2 + r
            val x = p.pos.x
            val z = p.pos.z
            if (!(abs(x) < hx && z > zMin && z < zMax)) continue
            val outLeft = x + hx
            val outRight = hx - x
            val outFront = if (dir > 0) zMax - z else z - zMin
            val outBack = if (dir > 0) z - zMin else zMax - z
            val least = Pitch.lesser(Pitch.lesser(outLeft, outRight), Pitch.lesser(outFront, outBack))
            p.pos = if (least == outLeft) {
                Vec(-hx, z)
            } else if (least == outRight) {
                Vec(hx, z)
            } else if (least == outFront) {
                Vec(x, if (dir > 0) zMax else zMin)
            } else {
                Vec(x, if (dir > 0) zMin else zMax)
            }
        }
    }
}
